File: talks/functional-programming/root.py
from math import log

# Find a zero of -x ln x - (1-x) ln (1-x) - x

def f(x):
    return (-x*math.log(x)
            - (1.0-x)*math.log(1.0-x) - x)

eps = 0.001

# Ensure f(a) > 0  and f(b) < 0
a = 0.5
b = 0.99

while abs(b - a) > eps:
    guess = (a + b) / 2.0
    if f(guess) > 0.0:
        a = guess
    else:
        b = guess
        
print(f"[{a},{b}]")

File: talks/functional-programming/talk1/root.hs
import Data.List (iterate, find)

f :: Float -> Float
f x = - x * log x - (1 - x) * log (1 - x) - x

narrowRange f (a, b) =
  let guess = (a + b) / 2.0 in
    if f guess > 0 then (guess, b) else (a, guess)

isNarrowed eps (a, b) = abs (b - a) < eps

findRoot eps a b f =
  find (isNarrowed eps) $ iterate (narrowRange f) (a, b)

main = print $ findRoot 0.001 0.5 0.99 f
